### ISSUE TITLE
Unable to pull images from private registry with Kubernetes and imagePullSecretes

### ISSUE THREAD

**Issue by bnason** [containerd/cri#848]
Kubelet is unabled to pull an image from a private registry when using `imagePullSecrets` in Kubernetes.

I've made the issue here instead of the Kubernetes repo because as far as I understand it, this works when using Docker.

Following the instructions at https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/

`kubectl create secret docker-registry regcred --docker-server=registry.example.com --docker-username=<username> --docker-password=<password> --docker-email=<email>`

```
apiVersion: v1
kind: Pod
metadata:
  name: private-reg
spec:
  containers:
  - name: private-reg-container
    image: registry.example.com/my/app:latest
  imagePullSecrets:
  - name: regcred
```

This is the error I receive
`Failed to pull image "registry.example.com/my/app:latest": rpc error: code = Unknown desc = failed to resolve image "registry.example.com/my/app:latest": failed to fetch anonymous token: unexpected status: 403 Forbidden`

I get the same error when I try to pull the image using ctr without specifying a --user

```
# ctr image pull registry.example.com/my/app:latest
ctr: failed to resolve reference "registry.example.com/my/app:latest": failed to fetch anonymous token: unexpected status: 403 Forbidden
```

Passing --user works though

```
# ctr image pull --user <user> registry.example.com/my/app:latest
Password:
registry.example.com/my/app:latest:                                                resolved       |++++++++++++++++++++++++++++++++++++++|
manifest-sha256:a20dd775dedf90b958f91f599cf02c06772c1e07d913bf9c62e6e826376caf86: exists         |++++++++++++++++++++++++++++++++++++++|
layer-sha256:adac4603423b99d0251a099b079cadc1868fc6c053e49cf6590604d6873c6a24:    exists         |++++++++++++++++++++++++++++++++++++++|
....
```

Here are the relevant kubelet logs

```
kubelet.go:1861] SyncLoop (ADD, "api"): "private-reg_web(b8ade9f1-85f7-11e8-9304-005056a6da01)"
reconciler.go:207] operationExecutor.VerifyControllerAttachedVolume started for volume "default-token-p2ccm" (UniqueName: "kubernetes.io/secret/b8ade9f1-85f7-11e8-9304-005056a6da01-default-token-p2ccm") pod "private-reg" (UID: "b8ade9f1-85f7-11e8-9304-005056a6da01")
reconciler.go:252] operationExecutor.MountVolume started for volume "default-token-p2ccm" (UniqueName: "kubernetes.io/secret/b8ade9f1-85f7-11e8-9304-005056a6da01-default-token-p2ccm") pod "private-reg" (UID: "b8ade9f1-85f7-11e8-9304-005056a6da01")
operation_generator.go:557] MountVolume.SetUp succeeded for volume "default-token-p2ccm" (UniqueName: "kubernetes.io/secret/b8ade9f1-85f7-11e8-9304-005056a6da01-default-token-p2ccm") pod "private-reg" (UID: "b8ade9f1-85f7-11e8-9304-005056a6da01")
kuberuntime_manager.go:385] No sandbox for pod "private-reg_web(b8ade9f1-85f7-11e8-9304-005056a6da01)" can be found. Need to start a new one
kubelet.go:1868] SyncLoop (UPDATE, "api"): "private-reg_web(b8ade9f1-85f7-11e8-9304-005056a6da01)"
remote_image.go:108] PullImage "registry.example.com/my/app:latest" from image service failed: rpc error: code = Unknown desc = failed to resolve image "registry.example.com/my/app:latest": failed to fetch anonymous token: unexpected status: 403 Forbidden
kuberuntime_image.go:51] Pull image "registry.example.com/my/app:latest" failed: rpc error: code = Unknown desc = failed to resolve image "registry.example.com/my/app:latest": failed to fetch anonymous token: unexpected status: 403 Forbidden
kuberuntime_manager.go:733] container start failed: ErrImagePull: rpc error: code = Unknown desc = failed to resolve image "registry.example.com/my/app:latest": failed to fetch anonymous token: unexpected status: 403 Forbidden
pod_workers.go:186] Error syncing pod b8ade9f1-85f7-11e8-9304-005056a6da01 ("private-reg_web(b8ade9f1-85f7-11e8-9304-005056a6da01)"), skipping: failed to "StartContainer" for "private-reg-container" with ErrImagePull: "rpc error: code = Unknown desc = failed to resolve image \"registry.example.com/my/app:latest\": failed to fetch anonymous token: unexpected status: 403 Forbidden"
kubelet.go:1906] SyncLoop (PLEG): "private-reg_web(b8ade9f1-85f7-11e8-9304-005056a6da01)", event: &pleg.PodLifecycleEvent{ID:"b8ade9f1-85f7-11e8-9304-005056a6da01", Type:"ContainerStarted", Data:"3c564fa72c7d503d7273e499f7a443907ffcb9ec2065f8aeb021e78fbcf746ca"}
kuberuntime_manager.go:513] Container {Name:private-reg-container Image:registry.example.com/my/app:latest Command:[] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:default-token-p2ccm ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] VolumeDevices:[] LivenessProbe:nil ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:Always SecurityContext:nil Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
pod_workers.go:186] Error syncing pod b8ade9f1-85f7-11e8-9304-005056a6da01 ("private-reg_web(b8ade9f1-85f7-11e8-9304-005056a6da01)"), skipping: failed to "StartContainer" for "private-reg-container" with ImagePullBackOff: "Back-off pulling image \"registry.example.com/my/app:latest\""
kuberuntime_manager.go:513] Container {Name:private-reg-container Image:registry.example.com/my/app:latest Command:[] Args:[] WorkingDir: Ports:[] EnvFrom:[] Env:[] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:default-token-p2ccm ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath: MountPropagation:<nil>}] VolumeDevices:[] LivenessProbe:nil ReadinessProbe:nil Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:Always SecurityContext:nil Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
pod_workers.go:186] Error syncing pod b8ade9f1-85f7-11e8-9304-005056a6da01 ("private-reg_web(b8ade9f1-85f7-11e8-9304-005056a6da01)"), skipping: failed to "StartContainer" for "private-reg-container" with ImagePullBackOff: "Back-off pulling image \"registry.example.com/my/app:latest\""

```

**Comment by Random-Liu** [comment]
@bnason Can you try `crictl pull`? If it does work, the problem should be in Kubernetes; if it doesn't work, then should be problem in containerd `cri` plugin. It helps us debug. Thanks!

**Comment by bnason** [comment]
@Random-Liu where can I find an up to date crictl binary?

**Comment by mikebrow** [comment]
it comes with the release tar files.. which version of containerd are you using.

**Comment by bnason** [comment]
```
# containerd --version
containerd github.com/containerd/containerd v1.1.1 d64c661f1d51c48782c9cec8fda7604785f93587
```

The tar ([containerd-1.1.1.linux-amd64.tar.gz](https://github.com/containerd/containerd/releases/download/v1.1.1/containerd-1.1.1.linux-amd64.tar.gz)) I downloaded for this version does not contain a crictl binary, just the following:
> bin/containerd
> bin/ctr
> bin/containerd-shim
> bin/containerd-release
> bin/containerd-stress

**Comment by bnason** [comment]
Here is the output of different attempts to pull the image.

```
# crictl pull registry.example.com/my/app:latest
DEBU[0000] PullImageRequest: &PullImageRequest{Image:&ImageSpec{Image:registry.example.com/my/app:latest,},Auth:nil,SandboxConfig:nil,}
DEBU[0000] PullImageResponse: nil
FATA[0000] pulling image failed: rpc error: code = Unknown desc = failed to resolve image "registry.example.com/my/app:latest": failed to fetch anonymous token: unexpected status: 403 Forbidden
```

```
# crictl pull --creds user registry.example.com/my/app:latest
DEBU[0000] PullImageRequest: &PullImageRequest{Image:&ImageSpec{Image:registry.example.com/my/app:latest,},Auth:&AuthConfig{Username:user,Password:,Auth:,ServerAddress:,IdentityToken:,RegistryToken:,},SandboxConfig:nil,}
DEBU[0000] PullImageResponse: nil
FATA[0000] pulling image failed: rpc error: code = Unknown desc = failed to resolve image "registry.example.com/my/app:latest": failed to fetch anonymous token: unexpected status: 403 Forbidden
```

```
# crictl pull --creds user:password registry.example.com/my/app:latest
DEBU[0000] PullImageRequest: &PullImageRequest{Image:&ImageSpec{Image:registry.example.com/my/app:latest,},Auth:&AuthConfig{Username:user,Password:password,Auth:,ServerAddress:,IdentityToken:,RegistryToken:,},SandboxConfig:nil,}
DEBU[0027] PullImageResponse: &PullImageResponse{ImageRef:sha256:50cf30d47283a47b5612eb7b984cce298d7950af0a2c3b0dc83119e20090db60,}
Image is up to date for sha256:50cf30d47283a47b5612eb7b984cce298d7950af0a2c3b0dc83119e20090db60
```

Does kubelet use different methods to pass credentials to Docker vs containerd?

**Comment by Random-Liu** [comment]
@bnason All these are with containerd?

**Comment by bnason** [comment]
@Random-Liu yes

**Comment by mikebrow** [comment]
> Does kubelet use different methods to pass credentials to Docker vs containerd?
It shouldn't be, docker and containerd are both integrating through kubernetes' CRI interface, which is what crictl uses.

**Comment by bnason** [comment]
Closing this as it was resolved in https://github.com/kubernetes/kubernetes/issues/66162
TLDR; docker-registry secret was created in 1.9 kubectl which created `dockercfg` secret. 1.10 (correct version of cluster) correctly created a `dockerconfigjson` file.

**Comment by mikebrow** [comment]
Thx for reporting and tracing it down @bnason. Cheers!

**Comment by shapeofarchitect** [comment]
Hi Folks , I am still facing this issue in my private registry and it just doesn't work. I have tried to use the exact syntax as mentioned above by @bnason  but to no avail it breaks at this.
Here is my containerd configuration , It would be of great help if there are any pointers to fix.

`[plugins]
  [plugins.cri.containerd]
    snapshotter = "overlayfs"
    [plugins.cri.containerd.default_runtime]
      runtime_type = "io.containerd.runtime.v1.linux"
      runtime_engine = "/usr/local/bin/runc"
      runtime_root = ""
    [plugins.cri.containerd.untrusted_workload_runtime]
      runtime_type = "io.containerd.runtime.v1.linux"
      runtime_engine = "/usr/local/bin/runsc"
      runtime_root = "/run/containerd/runsc"
    [plugins.cri.containerd.gvisor]
      runtime_type = "io.containerd.runtime.v1.linux"
      runtime_engine = "/usr/local/bin/runsc"
      runtime_root = "/run/containerd/runsc"
    [plugins.cri.cni]
      bin_dir = "/opt/cni/bin"
      conf_dir = "/etc/cni/net.d"
      conf_template = "/etc/cni/net.d/10-flannel.conflist"
    [plugins.cri.registry]
      [plugins.cri.registry.mirrors]
        [plugins.cri.registry.mirrors."mavenrepo.adaptivemobile.com:8085"]
          endpoint = ["https://mavenrepo.adaptivemobile.com:8085"]`

**Comment by sumitKash** [comment]
creating secrets with dockerconfig json resolves the issue. 
looks someting like this:
```
apiVersion: v1
data:
  .dockerconfigjson: "<base64 encoded credentials>
kind: Secret
metadata:
  name: <anyname>
  namespace: mailserver
type: kubernetes.io/dockerconfigjson
```
